add migration doc from wiki contents (#747)
canonical/ubuntu-server-documentation: docs/explanation/virtualisation/qemu-machine-types.md

---
myst:
  html_meta:
    description: Understand QEMU machine types on Ubuntu, and how to upgrade the machine type of an existing guest.
---

(qemu-machine-types)=
# QEMU Machine types

> The QEMU machine type defines the exact virtual hardware blueprint (chipset, bus architecture, and version-locked features) of a virtual machine. This allows QEMU to offer to virtual machines a well-defined and stable environment they can run not impacted by changes that happen under the hood like software upgrades or migrations to other systems.

Reviewer: ```suggestion
The QEMU machine type defines the exact virtual hardware blueprint (chipset, bus architecture, and version-locked features) of a virtual machine. This allows QEMU to offer VMs a well-defined and stable environment they can run without being impacted by changes that happen "under the hood", such as software upgrades or migrations to other systems.
```

Reviewer: do we want to link to the live migration page here?

> The QEMU machine type defines the exact virtual hardware blueprint (chipset, PCI slots, ACPI tables) of a virtual machine. This allows QEMU to offer to virtual machines a well-defined and stabilized foundation they can run on so that they are not impacted by changes that happen under the hood on the virtualisation software stack at machine upgrade or guests migration.

Reviewer 2: ```suggestion
The QEMU machine type defines the exact virtual hardware blueprint (chipset, bus architecture, and version-locked features) of a virtual machine. This allows QEMU to offer to virtual machines a well-defined and stable environment they can run not impacted by changes that happen under the hood like software upgrades or migrations to other systems.
```

The protection is to stay stable in behavior when code changes.
I also have been pondering about the list in the blueprint to make a proposal that (to me) seems better.

> However, when the QEMU software is upgraded, machine types can evolve and get new features that might be incompatible with the migrated virtual machine. The solution is to version the machine types. The following is the list of Q35 machine types available in QEMU 10.2:

Reviewer 2: ```suggestion
However, when the QEMU software is upgraded, it might get new features and behavior that would be incompatible with the current definition virtual machine. But since the purpose of these types is to lock the feature set, they can't be changed - the solution is to version the machine types. The following is the list of Q35 machine types available in QEMU 10.2:
```

Feel free to dislike it, but I think it is an improvement to the former

> However, when the QEMU software is upgraded, it might get new features and behavior that would be incompatible with the current definition virtual machine. But since the purpose of these types is to lock the feature set, they can't be changed - the solution is to version the machine types. The following is the list of Q35 machine types available in QEMU 10.2:

Reviewer: ```suggestion
However, when the QEMU software is upgraded, it might get new features and behavior that would be incompatible with the current definition of a given virtual machine. Since the purpose of these machine types is to lock the feature set, they can't be changed - so the solution is to version the machine types. The following is the list of Q35 machine types available in QEMU 10.2:
```

Reviewer: Unclear what Ubuntu versions this is available for. Also, do we want to provide a table here that we need to maintain? Is there a QEMU official page we can link to instead?

```text
...
pc-q35-8.0           Standard PC (Q35 + ICH9, 2009)
pc-q35-8.1           Standard PC (Q35 + ICH9, 2009)
pc-q35-8.2           Standard PC (Q35 + ICH9, 2009)
pc-q35-9.0           Standard PC (Q35 + ICH9, 2009)
pc-q35-9.1           Standard PC (Q35 + ICH9, 2009)
pc-q35-9.2           Standard PC (Q35 + ICH9, 2009)
pc-q35-10.0          Standard PC (Q35 + ICH9, 2009)
pc-q35-10.1          Standard PC (Q35 + ICH9, 2009)
pc-q35-10.2          Standard PC (Q35 + ICH9, 2009)
```

> ## Ubuntu specific machine types

Reviewer: ```suggestion
## Ubuntu-specific machine types
```

For each release, Ubuntu-specific machine types are added and point to the corresponding upstream machine types (see [Launchpad bug](https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/1294823)). On Ubuntu Resolute 26.04 LTS, for the x86 platforms, two new Ubuntu machine types are forked from the corresponding `i440fx` and `q35` upstream latest machine types:

```text
root@resolute:~# qemu-system-x86_64 -machine help | grep resolute
ubuntu               Ubuntu 26.04 PC (i440FX + PIIX, 1996) (alias of pc-i440fx-resolute)
pc-i440fx-resolute   Ubuntu 26.04 PC (i440FX + PIIX, 1996) (default)
ubuntu-q35           Ubuntu 26.04 PC (Q35 + ICH9, 2009) (alias of pc-q35-resolute)
pc-q35-resolute      Ubuntu 26.04 PC (Q35 + ICH9, 2009)
```

> `pc-q35-resolute` and `pc-i440fx-resolute` are forked respectively from `pc-q35-10.2` and `pc-i440fx-10.2` that are the latest x86 upstream machine types since QEMU 10.2 is the version for Ubuntu Resolute LTS. `pc-i440fx-resolute` is defined as the default machine type for every virtual machine and the `ubuntu` alias now points to it.

Reviewer: ```suggestion
`pc-q35-resolute` and `pc-i440fx-resolute` are forked respectively from `pc-q35-10.2` and `pc-i440fx-10.2`, which are the latest x86 upstream machine types, since QEMU 10.2 is the version for Ubuntu Resolute LTS. `pc-i440fx-resolute` is defined as the default machine type for every virtual machine and the `ubuntu` alias now points to it.
```

> These distro-specific machine types allow Ubuntu to support users who have deployed VMs on Ubuntu LTSes to be able to live migrate their VMs to the next LTS dealing with various changes we might introduce into an existing Ubuntu release through the SRU process. Whenever we need to make a change that modifies an existing machine type, we can make a "point-release" element in the Ubuntu names. This is not tied to a usual Ubuntu LTS point release, but to

Reviewer: "LTSes" is not a valid word - we should say "LTS releases".
Should we link to the live migration page here?

> These distro-specific machine types allow Ubuntu to support users who have deployed VMs on Ubuntu LTSes to be able to live migrate their VMs to the next LTS dealing with various changes we might introduce into an existing Ubuntu release through the SRU process. Whenever we need to make a change that modifies an existing machine type, we can make a "point-release" element in the Ubuntu names. This is not tied to a usual Ubuntu LTS point release, but to

Reviewer: ```suggestion
These distro-specific machine types allow Ubuntu to support users who have deployed VMs on Ubuntu LTSes to be able to live migrate their VMs to the next LTS, dealing with whatever changes may have been introduced between one LTS and the next. Whenever we need to make a change that modifies an existing machine type, we can make a "point-release" element in the Ubuntu names. This is not tied to the usual Ubuntu LTS point releases, but to
```

Reviewer: A bit confusingly worded overall

Author: your proposal changes (the part "may have been introduced between one LTS and the next") the meaning of the sentence and I think we want to keep the original meaning;

Reviewer: The original was quite confusingly worded though, I'm happy to keep the original meaning if we can find a way to make that more clear

> anything introducing a delta to the machine type / `vmstate`. Similar to CentOS/RHEL, we want:

Reviewer 2: ```suggestion
anything introducing a delta to the machine type / `vmstate`. The format generally follows:
```

```text
$MACHINE_TYPE-$DISTRO-RELEASE[-v*]
```

> This is an example of "point-release" machine types that have been added to address the [arch-capabilities bug](https://bugs.launchpad.net/bugs/2131822) in Noble:

Reviewer 2: ```suggestion
The following is an example of an update that required machine types versioning to track a change in behavior when addressing a bug in handling [arch-capabilities bug](https://bugs.launchpad.net/bugs/2131822) in Noble:
```

> This is an example of "point-release" machine types that have been added to address the [arch-capabilities bug](https://bugs.launchpad.net/bugs/2131822) in Noble:

Reviewer: ```suggestion
This is an example of "point-release" machine types that have been added to address the {lpbug}`arch-capabilities bug <2131822>` in Noble:
```

```text
pc-i440fx-questing-v2 Ubuntu 25.10 PC v2 (i440FX + PIIX, + 10.1 machine, 1996)
pc-i440fx-noble-v2   Ubuntu 24.04 PC v2 (i440FX + PIIX, arch-caps fix, 1996)
pc-q35-noble-v2      Ubuntu 24.04 PC v2 (Q35 + ICH9, arch-caps fix, 2009)
```

> ## Machine type handling summary

Reviewer 2: I'd ask you to drop all from "Machine types are handled by:" to "The default, if no machine type is specified, always points to the latest distribution-specific machine type."
You have explained all that above and doing so again is only redundancy and distracting.

After removal change the header "Machine type handling summary" => "Machine type availability"
The then remaining elements talk about why we keep upstream names and if/when we cleanup which fits the title.

Machine types are handled by:

- Adding a distribution-release-specific suffix to the default type(s)
  of each major architecture. Examples for Resolute:
  - x86: `pc-i440fx-resolute` and `pc-q35-resolute`
  - s390x: `s390-ccw-virtio-resolute`
  - ppc64el: `pseries-resolute`
  - arm64: `virt-resolute`
- Feature backports add a `-v%d` suffix to the affected types.
  - To avoid a proliferation of these types, such changes should be done carefully.
- The default, if no machine type is specified, always points to the latest
  distribution-specific machine type.
- We do not drop upstream types; they are provided as-is without further
  guarantees.
  - Cross-vendor/downstream migrations might work for upstream types, but are

>     considered unsupported. This has always been the case, but package documentation

Reviewer 2: We discussed that in 1:1, drop all of "This has always been the case, but package documentation might need updating to reflect this."

    might need updating to reflect this.

> - Cleanup matches the usual supported distribution upgrade paths.

Reviewer 2: We also discussed that Cleanup section to be silly nowadays and was an overlay aggressive past intention but never acted on, we had a better version that was a draft like:

```
=> The supported migration paths match the supported Ubuntu upgrade paths. So a release can always migrate to the subsequent release, and furthermore an LTS can migrate to the following LTS+1. In addition to that an LTS can be expected to also safely migrate to LTS+2. Other paths might work as well - in fact, many do - but only those matching upgrade paths are officially considered supported.
```

And we had this which is a nice "why do they even exist" to go where you explain the ubuntu themed names above:
```
=> types are for recognition as “ubuntu” and for avoiding the misbelief that they could neatly move elsewhere while we can’t give any guarantees. Furthermore redhat paved the way and they have distro named types as well
```

  - Drop former non-LTS release definitions after the next LTS.
  - Drop former LTS release definitions when out of support.

## Upgrading the machine type of your VM

> Upgrading the machine type of a virtual machine (VM) can be thought of in the same way as buying (virtual) hardware of the same spec but with a newer release date. Whereas to upgrade a physical machine you might buy an improved CPU, more RAM, or increased storage, with a virtual machine you can change the configuration to achieve the same results.

Reviewer 2: ```suggestion
Think about upgrading a physical machine:
* You might buy more CPUs and RAM, or increased storage - for a virtual machine you can to that by a change to its configuration.
* But you'd usually also buy a newer generation of board, chipset and so on - in a virtual machine the equivalent to that is using a newer machine type
```

I might have rewritten this too often, hope you like it

> ### Why should you do this for a VM?

Reviewer 2: Drop the whole "Why should you do this for a VM?" section please.
It is a lot "bla bla" and barely and holds up to a hard question. It could be super extended and then make sense, but isn't important to the story told. Remove it please. The updated suggested text above it making it clear what it is, people are used to think in real HW and that works IMHO much better.

### Backing up guest definitions

As usual, with major configuration changes it is wise to back up your guest definition and disk state to be able to do a rollback -- just in case something goes wrong.

### Upgrade the machine type

There is no integrated single command to update the machine type via `virsh` or similar tools. It is a normal part of your machine definition, and therefore updated the same way as most others.

#### Shut down the VM

First, shut down your machine and wait until it has reached that state:

```bash
virsh shutdown <your_machine>
```

You can check the status of the machine with the following command:

```{terminal}
:copy:
:user:
:host:
:dir:
virsh list --inactive
```

#### Edit the guest definition

Once the machine is listed as "shut off", you can then edit the machine definition and find the type in the `type` tag given at the machine attribute.

```bash
virsh edit <your_machine>

> <type arch='x86_64' machine='pc-i440fx-bionic'>hvm</type>

Reviewer 2: Nowadays virt-xml might do that more comfortably - see https://manpages.ubuntu.com/manpages/resolute/man1/virt-xml.1.html#boot and try if that works well for ubuntu types.

If it works see if the description would seem better with this and refer to the man page for more.

If it does not work that might be a patch that we need to track as bug anyway (low prio but valid).

Author: I am afraid that this command 

$ virt-xml --edit --boot refresh-machine-type=yes <domain>

will make the machine type to switch from Ubuntu machine type to upstream version type, for example:

pc-q35-noble -> pc-q35-11.0

```

Change this to the value you want. If you need to check what machine types are available via the `kvm -M ?` command first, then note that while upstream types are provided for convenience, only Ubuntu types are supported. There you can also see what the current default would be, as in this example:

```{terminal}
:copy:
:user:
:host:
:dir:
kvm -M ?

> pc-i440fx-xenial       Ubuntu 16.04 PC (i440FX + PIIX, 1996) (default)

Reviewer 2: While touching it - Maybe bump all the old examples to Noble+Resolute examples?
